### PR TITLE
Convert form-encoding body to model-based version.

### LINKED ISSUE #2064: Model API FormEncoding type.
We have `MultipartBuilder` and `FormEncodingBuilder` but they produce an opaque `RequestBody`, instead of a subclass that offers APIs to inspect the composite parts.

### PR DESCRIPTION
**Pull request by JakeWharton**
Update the Multipart version as well to match some conventions.

Closes #2064.

### CODE REVIEW

**Comment by JakeWharton** [comment]
This currently exposes the user-supplied String in the model API. We probably should behave more like HttpUrl.

File: okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import okio.Buffer;
+import okio.BufferedSink;
+
+import static com.squareup.okhttp.HttpUrl.FORM_ENCODE_SET;
+
+public final class FormEncodingBody extends RequestBody {
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: `Form` ?

File: okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  private static final MediaType CONTENT_TYPE =
+      MediaType.parse("application/x-www-form-urlencoded");
+
+  private final List<String> names;
+  private final List<String> values;
+  private final BitSet encoded;
+
+  public FormEncodingBody(List<String> names, List<String> values, BitSet encoded) {
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: private

File: okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  public List<String> names() {
+    return names;
+  }
+
+  public String name(int index) {
+    return names.get(index);
+  }
+
+  public List<String> values() {
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: We’ll probably want some map-like access also

Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: Is getting values as a list useful?

File: okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    this.values = Util.immutableList(values);
+    this.encoded = (BitSet) encoded.clone();
+  }
+
+  /** The number of key-value pairs in this form-encoded body. */
+  public int size() {
+    return names.size();
+  }
+
+  public List<String> names() {
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: Set?

Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: (I think of this as map keys)

File: okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    return names;
+  }
+
+  public String name(int index) {
+    return names.get(index);
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: this needs to decode?

File: okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  public List<String> values() {
+    return values;
+  }
+
+  public String value(int index) {
+    return values.get(index);
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: needs to decode?

File: okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java (hunk 3 above)
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: Maybe just omit this?

**Comment by swankjesse** [comment]
LGTM

I think we might want to follow-up with handy map-like access to the parameters

File: okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java (hunk 4 above)
Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: Nope. It's an ordered list of pairs. Repeated keys are valid!

Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/FormEncodingBody.java]: Although perhaps we want a Headers-like approach here.

**Comment by JakeWharton** [comment]
Renamed to `FormBody`, encode on the way in, JIT decode on the way out, expose encoded accessors.

**Pull request merged**